Java 16 incompatibility due to usage of internal APIs

The README promises that "Playwright requires Java 8 or newer". This does not include Java 16, however. With [JEP 396](https://openjdk.java.net/jeps/396), further APIs were encapsulated that are being used by Playwright. As a result, running the test suite with Java 16 fails. Examples:

```
[ERROR] com.microsoft.playwright.TestScreencast.shouldExposeVideoPath(Path)  Time elapsed: 0.096 s  <<< ERROR!
java.lang.reflect.InaccessibleObjectException: Unable to make field private final sun.nio.fs.UnixFileSystem sun.nio.fs.UnixPath.fs accessible: module java.base does not "opens sun.nio.fs" to unnamed module @152f2773
	at com.microsoft.playwright.TestScreencast.shouldExposeVideoPath(TestScreencast.java:30)

...


[ERROR] com.microsoft.playwright.TestBrowserContextAddCookies.shouldRoundtripCookie  Time elapsed: 0.103 s  <<< ERROR!
java.lang.reflect.InaccessibleObjectException: Unable to make private java.util.Collections$EmptyList() accessible: module java.base does not "opens java.util" to unnamed module @152f2773
	at com.microsoft.playwright.TestBrowserContextAddCookies.shouldRoundtripCookie(TestBrowserContextAddCookies.java:53)
```

Steps to reproduce:

1. Install any version of OpenJDK 16 (I'm using [Liberica by Bell Soft](https://bell-sw.com/pages/downloads/#/java-16-current))
2. Run `mvn test`

I know about workarounds like `--add-opens`. But it would be great if Playwright worked without them. Because the OpenJDK team plans to enable further restrictions in the future, it might make sense to test against the newest JDK as part of each CI run.

--- a/playwright/src/main/java/com/microsoft/playwright/impl/BrowserContextImpl.java
+++ b/playwright/src/main/java/com/microsoft/playwright/impl/BrowserContextImpl.java
@@ -19,13 +19,22 @@ package com.microsoft.playwright.impl;
 import com.google.gson.JsonArray;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
-import com.microsoft.playwright.*;
-import com.microsoft.playwright.options.*;
+import com.microsoft.playwright.BrowserContext;
+import com.microsoft.playwright.Page;
+import com.microsoft.playwright.PlaywrightException;
+import com.microsoft.playwright.Route;
+import com.microsoft.playwright.options.BindingCallback;
+import com.microsoft.playwright.options.Cookie;
+import com.microsoft.playwright.options.FunctionCallback;
+import com.microsoft.playwright.options.Geolocation;
 
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 import java.util.regex.Pattern;
@@ -35,7 +44,6 @@ import static com.microsoft.playwright.impl.Utils.isSafeCloseError;
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.nio.file.Files.readAllBytes;
 import static java.util.Arrays.asList;
-import static java.util.Collections.emptyList;
 
 class BrowserContextImpl extends ChannelOwner implements BrowserContext {
   private final BrowserImpl browser;
@@ -109,7 +117,7 @@ class BrowserContextImpl extends ChannelOwner implements BrowserContext {
 
   @Override
   public List<Cookie> cookies(String url) {
-    return cookies(url == null ? emptyList() : asList(url));
+    return cookies(url == null ? new ArrayList<>() : asList(url));
   }
 
   private void closeImpl() {
@@ -181,7 +189,7 @@ class BrowserContextImpl extends ChannelOwner implements BrowserContext {
   private List<Cookie> cookiesImpl(List<String> urls) {
     JsonObject params = new JsonObject();
     if (urls == null) {
-      urls = emptyList();
+      urls = new ArrayList<>();
     }
     params.add("urls", gson().toJsonTree(urls));
     JsonObject json = sendMessage("cookies", params).getAsJsonObject();
@@ -229,7 +237,7 @@ class BrowserContextImpl extends ChannelOwner implements BrowserContext {
       options = new GrantPermissionsOptions();
     }
     if (permissions == null) {
-      permissions = emptyList();
+      permissions = new ArrayList<>();
     }
     JsonObject params = gson().toJsonTree(options).getAsJsonObject();
     params.add("permissions", gson().toJsonTree(permissions));

--- a/playwright/src/main/java/com/microsoft/playwright/impl/Serialization.java
+++ b/playwright/src/main/java/com/microsoft/playwright/impl/Serialization.java
@@ -20,8 +20,9 @@ import com.google.gson.*;
 import com.google.gson.reflect.TypeToken;
 import com.google.gson.stream.JsonReader;
 import com.google.gson.stream.JsonWriter;
+import com.microsoft.playwright.ElementHandle;
+import com.microsoft.playwright.PlaywrightException;
 import com.microsoft.playwright.options.*;
-import com.microsoft.playwright.*;
 
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
@@ -51,7 +52,7 @@ class Serialization {
         .registerTypeHierarchyAdapter(JSHandleImpl.class, new HandleSerializer())
         .registerTypeAdapter((new TypeToken<Map<String, String>>(){}).getType(), new StringMapSerializer())
         .registerTypeAdapter((new TypeToken<Map<String, Object>>(){}).getType(), new FirefoxUserPrefsSerializer())
-        .registerTypeAdapter(Path.class, new PathSerializer()).create();
+        .registerTypeHierarchyAdapter(Path.class, new PathSerializer()).create();
     }
     return gson;
   }

--- a/playwright/src/main/java/com/microsoft/playwright/impl/Utils.java
+++ b/playwright/src/main/java/com/microsoft/playwright/impl/Utils.java
@@ -16,30 +16,41 @@
 
 package com.microsoft.playwright.impl;
 
-import com.google.gson.Gson;
+import com.google.gson.*;
 import com.microsoft.playwright.PlaywrightException;
 import com.microsoft.playwright.options.FilePayload;
 
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.reflect.Type;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.*;
 
 class Utils {
   // TODO: generate converter.
   static <F, T> T convertViaJson(F f, Class<T> t) {
-    Gson gson = new Gson();
+    Gson gson = new GsonBuilder()
+      // Necessary to avoid access to private fields/classes,
+      // see https://github.com/microsoft/playwright-java/issues/423
+      .registerTypeAdapter(Optional.class, new OptionalSerializer())
+      .create();
     String json = gson.toJson(f);
+    System.err.println("json = " + json);
     return gson.fromJson(json, t);
   }
 
-  static boolean isFunctionBody(String expression) {
-    expression = expression.trim();
-    return expression.startsWith("function") ||
-      expression.startsWith("async ") ||
-      expression.contains("=>");
+  private static class OptionalSerializer implements JsonSerializer<Optional> {
+    @Override
+    public JsonElement serialize(Optional src, Type typeOfSrc, JsonSerializationContext context) {
+      JsonObject result = new JsonObject();
+      if (src.isPresent()) {
+        result.add("value", context.serialize(src.get()));
+      }
+      return result;
+    }
   }
 
   static Set<Character> escapeGlobChars = new HashSet<>(Arrays.asList('/', '$', '^', '+', '.', '(', ')', '=', '!', '|'));

